### PR TITLE
Move to Platform team's Find All References Kind Column

### PR DESCRIPTION
**Pull request by mavasani**
Delete a bunch of code on Roslyn side, which is no longer necessary as we don't own the Kind column anymore, it has been moved to the platform layer and shared with other teams.

I have also cleaned up some code in this space for newly added ContainingType and ContainingMember columns.

Fixes VSO bug [#822324](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/822324)

### CODE REVIEW

File: src/Features/Core/Portable/FindUsages/DefinitionItem.cs [hunk 1]
```diff
@@ -127,7 +126,7 @@ internal abstract partial class DefinitionItem
             ImmutableArray<TaggedText> originationParts,
             ImmutableArray<DocumentSpan> sourceSpans,
             ImmutableDictionary<string, string> properties,
-            ImmutableArray<FindUsageProperty> displayableProperties,
+            ImmutableDictionary<string, string> displayableProperties,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: should this be a multi-dictionary?  in the case of something like 'reference kind' i thought we could have multiple values associated with the key.

**Comment by CyrusNajmabadi** [comment]
> Delete a bunch of code on Roslyn side, which is no longer necessary as we don't own the Kind column anymore, it has been moved to the platform layer and shared with other teams.

> I have also cleaned up some code in this space for newly added ContainingType and ContainingMember columns.

Any chance this could be two PRs?

File: src/Features/Core/Portable/FindUsages/DefinitionItem.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: We don't have a reference kind for `DefinitionItem`. `SymbolUsageInfo` is now a directly property on `SourceReferenceItem`

**Comment by mavasani** [comment]
> Any chance this could be two PRs?

Actually, most of this PR is just deleting bunch of code. Deleting the code for Kind column opens up scope to perform more cleanup for code around ContainingType and ContainingMember columns. So it seemed more logical to do it in one go.

File: src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj [hunk 2]
```diff
@@ -201,6 +201,7 @@
     <PackageReference Include="NuGet.VisualStudio" Version="$(NuGetVisualStudioVersion)" />
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonVersion)" />
     <PackageReference Include="Microsoft.VisualStudio.SDK.EmbedInteropTypes" Version="$(MicrosoftVisualStudioSDKEmbedInteropTypesVersion)" />
+    <PackageReference Include="System.Threading.Tasks.Dataflow" Version="$(SystemThreadingTasksDataflowVersion)" />
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: why?

**Comment by CyrusNajmabadi** [comment]
Could you point out where in the PR we're now using their column? thanks!

File: src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj (hunk 2 above)
Comment (mavasani) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: Moving to newer Shell packages seems to cause MSBuild version conflicts from implicitly chosen version and version coming from Shell packages. Few more are still there: https://dev.azure.com/dnceng/public/_build/results?buildId=367711&view=logs

```
Build FAILED.

F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\TestUtilities\Roslyn.Services.Test.Utilities.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\TestUtilities\Roslyn.Services.Test.Utilities.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Test\Diagnostics\Roslyn.Hosting.Diagnostics.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\CSharpTest2\Microsoft.CodeAnalysis.CSharp.EditorFeatures2.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\CSharpTest\Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\CSharpTest\Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Tools\IdeBenchmarks\IdeBenchmarks.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Tools\IdeBenchmarks\IdeBenchmarks.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\Test\Microsoft.CodeAnalysis.EditorFeatures.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Interactive\HostTest\InteractiveHost.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\VisualBasicTest\Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests.vbproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\EditorFeatures\VisualBasicTest\Microsoft.CodeAnalysis.VisualBasic.EditorFeatures.UnitTests.vbproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Features\LanguageServer\ProtocolUnitTests\Microsoft.CodeAnalysis.LanguageServer.Protocol.UnitTests.csproj]
F:\workspace.2\_work\1\s\.tools\msbuild\16.0.0-alpha\tools\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(2114,5): error MSB3277: Found conflicts between different versions of "System.Threading.Tasks.Dataflow" that could not be resolved.  These reference conflicts are listed in the build log when log verbosity is set to detailed. [F:\workspace.2\_work\1\s\src\Features\LanguageServer\ProtocolUnitTests\Microsoft.CodeAnalysis.LanguageServer.Protocol.UnitTests.csproj]
```

**Comment by mavasani** [comment]
> Could you point out where in the PR we're now using their column? thanks!

Just search for ".SymbolKind" in the files changed view - there should be 3 instances.

File: src/Features/Core/Portable/FindUsages/DefinitionItem.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: let me rephrase:  we already have the concept of "key => multiple vlaues".  so i would have expected the displayableProperties to follow that as well, and not be a `string => string`.

File: src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj (hunk 2 above)
Comment (mavasani) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: @dotnet/roslyn-infrastructure for guidance. There is probably a better way to resovle these conflicts.

**Comment by CyrusNajmabadi** [comment]
Ok.  So SymbolReferenceKinds is a new platform enum?

**Comment by mavasani** [comment]
> Ok. So SymbolReferenceKinds is a new platform enum?

Yes.

File: src/Features/Core/Portable/FindUsages/DefinitionItem.cs (hunk 1 above)
Comment (mavasani) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: > we already have the concept of "key => multiple vlaues".

Where do we have that concept? We used to have it in SourceReferenceItem, but I have removed that as part of this PR. Now the extensible custom properties for SourceReferenceItem are `string => string`, with an explicit value stored for SymbolUsageInfo. That helped cleanup and simplify bunch of code as we don't need `key => multiple vlaues` anymore.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: sorry, i'm not bein clear :)

Say i'm a lanuage plugging into FindRefs.  And i want to add a custom column.  *And i would like it to behave like the 'kind' column*.  There appears to be no way for me to do that.  Yes, i could add a column with singular string vlues.  But i could not add a column with multi-values.

This effectively means that hte extensibility system is less capable than what's built in.  :)

Comment (mavasani) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: Sure, that was our goal originally, but none of the concrete features require that currently and was causing unnecessary complexity in code. Once we have an actual customer requesting multivalue scenario, we can refactor the code to support it.

File: src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj (hunk 2 above)
Comment (JoeRobich) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: Hmm.. We've been seeing this issue when validating our 16.3 NuGet packages on .NET Framework projects. In that case adding the "Microsoft.VisualStudio.LanguageServices" package to the test project adds the 4.5.24 version of the Tasks.Dataflow package whereas VS seems to be building against 4.6.3 and build warnings are given.

Comment (mavasani) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: >  adds the 4.5.24 version of the Tasks.Dataflow package whereas VS seems to be building against 4.6.3 and build warnings are given.

Those are the exact conflicts I have been seeing. If you just try to apply the changes to `Versions.props` in this PR on top of master, you will see these MSBuild warnings locally on building Microsoft.VisualStudio.LanguageServices

Comment (JoeRobich) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: I've reached out to get some guidance on this. For the time being I think this is ok and hopefully solves both of our issues.

Comment (mavasani) [reply src/VisualStudio/Core/Def/Microsoft.VisualStudio.LanguageServices.csproj]: Thanks. Note that I would need to add this package reference to all the test projects mentioned in https://github.com/dotnet/roslyn/pull/38894#discussion_r328855499 as well.

File: src/Features/Core/Portable/FindUsages/DefinitionItem.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: i'm ok with that :)    This was more about me thinkin about this as if it was a public APi.  in that case, my preference would be to support this from the start.  But since this is all internal, i think this is 100% fine for now :)

Comment (mavasani) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: One tricky thing with multiple column values is how to handle resource strings to display when there are different combinations of values based on column filtering. Default way of comma based join might not always suffice. Best to avoid adding this support until we have a real scenario that can be tested.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/FindUsages/DefinitionItem.cs]: My ideal would be that hte runtime would allow you to specify an `object[]` (or `string[]`) and they'd just take care of it.  But i agree there's no pressing need to figure this out currently :)

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs [hunk 3]
```diff
@@ -704,84 +702,80 @@ private static bool IsNodeOrAnyAncestorLeftSideOfDot(SyntaxNode node, ISyntaxFac
...
                 containingSymbol.GetMemberType() != null &&
-                TryGetAdditionalProperty(ContainingMemberInfoPropertyName, containingSymbol, out var findUsagePropertyForContainingMember))
+                TryGetAdditionalProperty(ContainingMemberInfoPropertyName, containingSymbol, out var containingMemberProperty))
             {
-                additionalProperties.Add((FindUsageProperty)findUsagePropertyForContainingMember);
+                additionalProperties.Add(containingMemberProperty);
             }
 
-            return additionalProperties.ToImmutableAndFree();
+            return additionalProperties.ToImmutable();
```
Comment (JoeRobich) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs]: Why the change from `.ToImmutableAndFree` to `.ToImmutable`?

Comment (mavasani) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder.cs]: I switched from using a pooled ArrayBuilder to ImmutableDictionary.Builder as we now return a dictionary instead of an array of key-value pairs.
